File: programmers/etc/level2/빛의_경로_사이클.py
dy, dx = (1, 0, -1, 0), (0, -1, 0, 1)

def solution(grid):
    answer = []
    N, M = len(grid), len(grid[0])
    visited = [[[False]*4 for _ in range(M)] for _ in range(N)]
    for y in range(N):
        for x in range(M):
            for d in range(4):
                if visited[y][x][d]: continue
                cnt = 0
                ny, nx = y, x
                while not visited[ny][nx][d]:
                    visited[ny][nx][d] = True
                    cnt += 1
                    if grid[ny][nx] == "S":
                        pass
                    elif grid[ny][nx] == "L":
                        d = (d-1)%4
                    elif grid[ny][nx] == "R":
                        d = (d+1)%4
                    ny, nx = (ny+dy[d])%N, (nx+dx[d])%M
                answer.append(cnt)
    return sorted(answer)


grid = ["SL","LR"]
print(solution(grid)) # [16]

grid = ["S"]
print(solution(grid)) # [1,1,1,1]

grid = ["R","R"]
print(solution(grid)) # [4,4]

File: Java/programmers/etc/level2/빛의경로사이클.java
package CodeTest.Java.programmers.etc.level2;

import java.util.*;

public class 빛의경로사이클 {
    static int[] dy = {-1,0,1,0};
    static int[] dx = {0,-1,0,1};

    public List<Integer> solution(String[] grid) {
        List<Integer> answer = new ArrayList<>();
        int N = grid.length, M = grid[0].length();
        boolean[][][] visited = new boolean[N][M][4];

        for (int y=0; y<N; y++) {
            for (int x=0; x<M; x++) {
                for (int d=0; d<4; d++) {
                    if (visited[y][x][d]) continue;
                    int cnt = 0;
                    int ny = y, nx = x;
                    while (!visited[ny][nx][d]) {
                        visited[ny][nx][d] = true;
                        cnt++;
                        if (grid[ny].charAt(nx) == 'L') d = (d+3)%4;
                        if (grid[ny].charAt(nx) == 'R') d = (d+1)%4;
                        ny = (ny+dy[d]+N)%N; nx = (nx+dx[d]+M)%M;
                    }
                    answer.add(cnt);
                }
            }
        }

        Collections.sort(answer);
        return answer;
    }
}
